### PR TITLE
Add Taproot transaction feature tooltip with fee saving information

### LINKED ISSUE #2107: Add Taproot to transaction features menu
Hey guys,

To help with spreading awareness, and hopefully adoption, of bech32m (P2TR), Mempool could add a Taproot option in the features section when viewing a transaction. When hovering over it you could highlight the fee savings and potential privacy benefits this transaction could have had if using this address type. 

Highlighted the section below:

image

### PR DESCRIPTION
**Pull request by antonilol**
fixes #2107

also, fixes a typo in rbf tooltip and adds fee saving tooltips with p2pk and pay to multisig (bare multisig) inputs

~reviewers: the diff of frontend/src/app/components/address-labels/address-labels.component.ts is complete trash because the indentation changed.~ i moved the function out of the class and exported it to use it somewhere else

`realizedGains` has been renamed to `realizedBech32Gains`

### CODE REVIEW

**Comment by antonilol** [comment]
got a few transactions to test:

testnet:
some p2wshs: f03d001b7f5a9f2ed2be1dc19d9f8d42637726e76fb38da4dab025f5e247639f
bare ms: 7193e8e9894069253b52ebcdc74caf8eed5d93c00022dcf23fa096b27180fe21

mainnet:
p2pkh inputs: 4d8a9573f7f8b56940ff14026d4d4ce033c60e94290da0dfbb82745e5f62a050
the first non-coinbase tx (p2pk): f4184fc596403b9d638783cf57adfe4c75c605f6356fbc91338530e9831e9e16
taproot and 2of2s: 0b3cdeb16cffdf440176d6cf22d717ad72bbf04bf345f3c95021a54f8bc63d3b

**Comment by softsimon** [comment]
Good work @antonilol 

We deployed this PR on https://mempool.ninja/

@Xekyo @Bosch-0 would you mind take a look.

File: frontend/src/app/components/address-labels/address-labels.component.ts [hunk 1]
```diff
@@ -98,44 +98,52 @@ export class AddressLabelsComponent implements OnChanges {
...
-    const opM = ops.pop();
-    if (!opM.startsWith('OP_PUSHNUM_')) {
+  }
+
+  handleVout() {
+    this.detectMultisig(this.vout.scriptpubkey_asm);
+  }
+}
+
+export function parseMultisigScript(script: string): void | { m: number, n: number } {
```
Comment (softsimon) [inline frontend/src/app/components/address-labels/address-labels.component.ts:113]: Can you move this method into the `bitcoin.utils.ts` file instead.

**Review by softsimon — CHANGES_REQUESTED** [review]

File: frontend/src/app/components/address-labels/address-labels.component.ts (hunk 1 above)
Comment (antonilol) [reply frontend/src/app/components/address-labels/address-labels.component.ts:113]: ok yeah, also looks cleaner not importing from a component

File: frontend/src/app/bitcoin.utils.ts [hunk 2]
```diff
@@ -1,68 +1,117 @@
...
         const fullGains = scriptSigSize(vin) * 3;
         potentialBech32Gains += fullGains;
         potentialP2shGains += fullGains - (isP2pkh ? P2SH_P2WPKH_COST : P2SH_P2WSH_COST);
         break;
+      }
+    }
 
-    // TODO: should we also consider P2PK and pay-to-bare-script (non-p2sh-wrapped) as upgradable to P2WPKH and P2WSH?
+    if (isP2tr) {
+      if (vin.witness.length == 1) {
```
Comment (softsimon) [inline frontend/src/app/bitcoin.utils.ts:68]: Use tripple equals === when possible

Comment (softsimon) [inline frontend/src/app/bitcoin.utils.ts:68]:
```suggestion
      if (vin.witness.length === 1) {
```

**Review by wiz — CHANGES_REQUESTED** [review]
Tested NACK @ v2.5.0-dev [[777a1bb4](https://github.com/mempool/mempool/commit/777a1bb4)]

### Test case
https://mempool.ninja/tx/2eb8dbaa346d4be4e82fe444c2f0be00654d8cfd8c4a9a61b11aeaab8c00b272

### Expected Result

Screen Shot 2022-07-24 at 18 40 16

### Actual result

Screen Shot 2022-07-24 at 18 40 20

**Comment by antonilol** [comment]
> Tested NACK @ v2.5.0-dev [[777a1bb4](https://github.com/mempool/mempool/commit/777a1bb4)]
> 
> ### Test case
> https://mempool.ninja/tx/2eb8dbaa346d4be4e82fe444c2f0be00654d8cfd8c4a9a61b11aeaab8c00b272

hmmm yeah, thought this would happen. it doesnt show the label because all taproot inputs use tapscript which costs more fees than p2wsh...
we could compare to p2sh to always be able to show how much fees have been saved (and more)

**Comment by antonilol** [comment]
@wiz with 46bce30 the badge should show up on 2eb8dbaa346d4be4e82fe444c2f0be00654d8cfd8c4a9a61b11aeaab8c00b272, but because there are no fees saved compared to p2wsh the fee saving part is left out

**Review by wiz — APPROVED** [review]
Tested ACK @ v2.5.0-dev [[46bce30a](https://github.com/mempool/mempool/commit/46bce30a)]

**Review by softsimon — APPROVED** [review]
tested ACK

**Comment by wiz** [comment]
@xekyo we've merged this for now but would appreciate your post-merge review to ensure the math is correct :sweat_smile:

**Review by Xekyo — COMMENTED** [review]
Sorry, I'm not very familiar with the codebase. I hope my comments make sense.

It seems a bit strange to me to define the savings in terms of the actually appearing witness script for single-sig, since all the weights for single-sig are known anyway. I find it a bit hard to find out what the expected outcome would be, especially since "potential savings" seems to use two different meanings. 

I would like to suggest that some test cases are added, as it would then be much easier to compare the expected outcome with the product of the code.

File: frontend/src/app/components/tx-features/tx-features.component.html [hunk 3]
```diff
@@ -6,11 +6,14 @@
   </ng-template>
 </ng-template>
 
-<span *ngIf="segwitGains.realizedTaprootGains && !segwitGains.potentialTaprootGains; else notFullyTaproot" class="badge badge-success mr-1" i18n-ngbTooltip="Tooltip about privacy and fees saved with taproot" ngbTooltip="This transaction uses Taproot and thereby increased privacy and saved at least {{ segwitGains.realizedTaprootGains * 100 | number: '1.0-0' }}% on fees" placement="bottom" i18n="tx-features.tag.taproot|Taproot">Taproot</span>
+<span *ngIf="segwitGains.realizedTaprootGains && !segwitGains.potentialTaprootGains; else notFullyTaproot" class="badge badge-success mr-1" i18n-ngbTooltip="Tooltip about privacy and fees saved with taproot" ngbTooltip="This transaction uses Taproot and the user's thereby increased privacy and saved at least {{ segwitGains.realizedTaprootGains * 100 | number: '1.0-0' }}% on fees" placement="bottom" i18n="tx-features.tag.taproot|Taproot">Taproot</span>
```
[the remark this answers is not in the document]
Comment (antonilol) [reply frontend/src/app/components/tx-features/tx-features.component.html:9]: i pasted it in the wrong place, must be after "increased" (consistency with the other tooltips)

File: frontend/src/app/bitcoin.utils.ts [hunk 4]
```diff
@@ -1,68 +1,117 @@
...
+        // not single sig, not multisig: the complex scripts
+        // rough calculations on spending paths
+        // every OP_IF and OP_NOTIF indicates an _extra_ spending path, so add 1
+        const spendingPaths = script.split(' ').filter(op => /^(OP_IF|OP_NOTIF)$/g.test(op)).length + 1;
+        // now assume the script could have been split in ${spendingPaths} equal tapleaves
+        replacementSize = script.length / 2 / spendingPaths +
+        // but account for the leaf and branch hashes and internal key in the control block
+          32 * Math.log2((spendingPaths - 1) || 1) + 33;
+      }
+      potentialTaprootGains += witnessSize(vin) + scriptSigSize(vin) * 4 - replacementSize;
```
[the remark this answers is not in the document]
Comment (antonilol) [reply frontend/src/app/bitcoin.utils.ts:102]: i dont know if i understand what you mean, but the segwit tooltip has an "in between" option (wrapped segwit) which taproot doesnt have. one input can have non zero potentialBech32Gains and realizedBech32Gains (p2sh-p2w{s,pk}h), but one input can only have non zero potentialTaprootGains (non taproot) or non zero realizedTaprootGains (taproot)

[the remark this answers is not in the document]
Comment (Xekyo) [reply frontend/src/app/bitcoin.utils.ts:102]: I guess it wasn't clear to me how you were ordering the output types in the hierarchy. I would have expected P2TR to just be added after P2WPKH for input savings, but it was implemented as a separate comparison altogether. All good, sorry for the confusion.

File: frontend/src/app/bitcoin.utils.ts [hunk 5]
```diff
@@ -5,64 +5,151 @@ const P2SH_P2WSH_COST  = 35 * 4; // the WU cost for the non-witness part of P2SH
...
+    const isP2pkh        = vin.prevout.scriptpubkey_type === 'p2pkh';
+    const isP2sh         = vin.prevout.scriptpubkey_type === 'p2sh';
+    const isP2wsh        = vin.prevout.scriptpubkey_type === 'v0_p2wsh';
+    const isP2wpkh       = vin.prevout.scriptpubkey_type === 'v0_p2wpkh';
+    const isP2tr         = vin.prevout.scriptpubkey_type === 'v1_p2tr';
 
     const op = vin.scriptsig ? vin.scriptsig_asm.split(' ')[0] : null;
     const isP2sh2Wpkh = isP2sh && !!vin.witness && op === 'OP_PUSHBYTES_22';
-    const isP2sh2Wsh = isP2sh && !!vin.witness && op === 'OP_PUSHBYTES_34';
+    const isP2sh2Wsh  = isP2sh && !!vin.witness && op === 'OP_PUSHBYTES_34';
```
[the remark this answers is not in the document]
Comment (chadmartinez22) [reply frontend/src/app/bitcoin.utils.ts:28]: 😜😘

**Pull request merged**